Rework web client into a struct that can be accessed by JS

sisou opened the pull request:
## What's in this pull request?
Change the web client to be a struct, which is represented as a class in Javascript, exposing it's methods to be called from JS. For now it only implements subscription methods, which then use separately exposed-from-JS global listeners to communicate the events back to JS. But it can now be easily extended to provide active methods that return data to JS, or send signed transactions to the network, etc.

Interestingly, since the created `Client` is stored in Rust's heap (or stack, I can never remember which is which) when it's made available to JS with `wasm_bindgen`, it is never dereferenced and lives forever (technically only until `.free()` is called on it from JS-side, which is never). So we do not need the statistics-interval-hack anymore to keep the client in-scope =).

> **Note for reviewers:** It is recommended to hide whitespace changes when reviewing, as most existing code in `lib.rs` has simply been indented by one level.

## Pull request checklist

- [x] All tests pass. The project builds and runs.
- [x] I have resolved any merge conflicts.
- [x] I have resolved all `clippy` and `rustfmt` warnings.

sisou committed:
Rework web client into a struct that can be accessed by JS

jsdanielh commented on web-client/Cargo.toml:
```diff
@@ -21,12 +21,18 @@ maintenance = { status = "experimental" }
 wasm-bindgen = "0.2"
 wasm-bindgen-futures = "0.4"
 
+nimiq-hash = { path = "../hash" }
+nimiq-blockchain-interface = { path = "../blockchain-interface" }
+nimiq-blockchain-proxy = { path = "../blockchain-proxy", default-features = false }
+nimiq-consensus = { path = "../consensus", default-features = false }
+nimiq-network-interface = { path = "../network-interface" }
 nimiq-network-libp2p = { path = "../network-libp2p" }
```
We try to keep them alphabetically ordered:
```suggestion
beserial = { path = "../beserial", features = ["derive"] }
nimiq-blockchain-interface = { path = "../blockchain-interface" }
nimiq-blockchain-proxy = { path = "../blockchain-proxy", default-features = false }
nimiq-consensus = { path = "../consensus", default-features = false }
nimiq-hash = { path = "../hash" }
nimiq-network-interface = { path = "../network-interface" }
nimiq-network-libp2p = { path = "../network-libp2p" }
```

jsdanielh commented on web-client/Cargo.toml:
```diff
@@ -21,12 +21,18 @@ maintenance = { status = "experimental" }
 crate-type = ["cdylib"]
 
 [dependencies]
+beserial = { path = "../beserial", features = ["derive"] }
```
This is actually an internal crate, we regularly try to list them apart (see suggestion below)

jsdanielh commented on web-client/src/lib.rs:
```diff
@@ -15,101 +16,211 @@ pub use nimiq::{
     error::Error,
     extras::{panic::initialize_panic_reporting, web_logging::initialize_web_logging},
 };
+use nimiq_hash::Blake2bHash;
 
+use nimiq_blockchain_interface::{AbstractBlockchain, BlockchainEvent};
+use nimiq_blockchain_proxy::BlockchainProxy;
+use nimiq_consensus::ConsensusEvent;
+use nimiq_network_interface::network::{Network, NetworkEvent};
 use nimiq_network_libp2p::Multiaddr;
```
```suggestion

use beserial::Serialize;
use nimiq_blockchain_interface::{AbstractBlockchain, BlockchainEvent};
use nimiq_blockchain_proxy::BlockchainProxy;
use nimiq_consensus::ConsensusEvent;
use nimiq_hash::Blake2bHash;
use nimiq_network_interface::network::{Network, NetworkEvent};
use nimiq_network_libp2p::Multiaddr;
```

jsdanielh commented on web-client/src/lib.rs:
```diff
@@ -1,5 +1,6 @@
 use std::str::FromStr;
 
+use beserial::Serialize;
```
`beserial` is an internal crate, we try to list it apart, see suggestion below

jsdanielh commented on web-client/src/lib.rs:
```diff
@@ -15,101 +16,211 @@ pub use nimiq::{
+            }),
+        )
+        .await
+        .expect("Client initialization failed ");
```
```suggestion
        .expect("Client initialization failed");
```

jsdanielh commented on web-client/src/lib.rs:
```diff
@@ -15,101 +16,211 @@ pub use nimiq::{
+        };
+        config.network.seeds = vec![seed];
+
+        log::debug!("Final configuration: {:#?}", config);
```
We should change this to `tracing style`:
```suggestion
        log::debug!(?config, "Final configuration");
```

jsdanielh commented on web-client/src/lib.rs:
```diff
@@ -15,101 +16,211 @@ pub use nimiq::{
+        }
+    }
+
+    pub async fn subscribe_peers(&self) {
```
I'd suggest to add rustdoc documentation

jsdanielh commented on web-client/src/lib.rs:
```diff
@@ -15,101 +16,211 @@ pub use nimiq::{
 
-    // Run periodically
-    let interval = IntervalStream::new(statistics_interval as u32 * 1000);
+    pub async fn subscribe_blocks(&self) {
```
I'd suggest to add rustdoc documentation

jsdanielh commented on web-client/src/lib.rs:
```diff
@@ -15,101 +16,211 @@ pub use nimiq::{
-        statistics_interval = 10;
-        show_statistics = false;
+#[wasm_bindgen]
+pub struct WebClient {
```
I'd suggest to add rustdoc documentation

jsdanielh commented on web-client/src/lib.rs:
```diff
@@ -15,101 +16,211 @@ pub use nimiq::{
+
+#[wasm_bindgen]
+impl WebClient {
+    pub async fn launch() -> WebClient {
```
I'd suggest to add rustdoc documentation

sisou committed:
Add rustdoc documentation blocks

sisou commented:
Thank you, @jsdanielh! I overlooked the dependency ordering. Fixed that, plus the other comments. I also renamed `WebClient.launch()` to `WebClient.create()`, as I think it better represents what's happening and discourages calling it twice (which would create two competing WebClient instances).
